nio/client/async_client.py

```diff
@@ -3160,3 +3174,131 @@ async def set_pushrule_actions(
+            initial_state=new_room_initial_state,
+            predecessor={
+                "event_id": old_room_last_event.event_id,
+                "room_id": old_room_id
+            })
+
+        if isinstance(new_room, RoomCreateError):
+            return RoomUpgradeError("Room creation failed")
+
+        # Send tombstone event to the old room
+        old_room_tombstone = await self.room_put_state(old_room_id, "m.room.tombstone",
+                                                       {"body": room_upgrade_message,
+                                                        "replacement_room": new_room.room_id
+                                                       })
+        if isinstance(old_room_tombstone, RoomPutStateError):
+            return RoomUpgradeError("Failed to put m.room.tombstone")
+
+        # Remove old room from the room directory
+        old_room_alias = await self.room_get_state_event(old_room_id, "m.room.canonical_alias")
+        if isinstance(old_room_alias, RoomGetStateEventResponse):
+            aliases = list()
+            aliases.append(old_room_alias.content['alias'])
+            if 'alt_aliases' in old_room_alias.content:
+                alt_aliases = old_room_alias.content['alt_aliases']
+                for alias in alt_aliases:
+                    aliases.append(alias)
+            else:
+                alt_aliases = []
+
+            # Remove old room aliases
+            old_room_put_alias = await self.room_put_state(old_room_id,
+                                                           "m.room.canonical_alias", {
+                                                               "alt_aliases": []
+                                                           })
+
+            if isinstance(old_room_put_alias, RoomPutStateError):
+                return RoomUpgradeError("Failed to put m.room.canonical_alias")
+
+            for alias in aliases:
+                # Delete old aliases from room dir
+                method, path = Api.delete_room_alias(self.access_token, alias)
+                await self._send(DeleteAliasResponse, method, path)
+
+                method, path, data = Api.put_room_alias(self.access_token, alias, new_room.room_id)
+                await self._send(PutAliasResponse, method, path, data)
+
+            # Add new aliases to the new room
+            new_room_put_alias = await self.room_put_state(new_room.room_id,
+                                                           "m.room.canonical_alias",
+                                                           {
+                                                              "alias": old_room_alias.content['alias'],
+                                                              "alt_aliases": alt_aliases
+                                                           })
+            if isinstance(new_room_put_alias, RoomPutStateError):
+                return RoomUpgradeError("Failed to put m.room.canonical_alias")
```

Comment:
Oh, that's not even what I meant :D

I mean that splitting out the whole procedure of "move all local aliases from one room to another" would be good here, because that's a fairly long bit. Splitting out the reusable things of setting and removing aliases is also good of course.

Reply:
I have implemented a helper method for room alias management, it does not move aliases tho, at this is nothing that happens that often and can be done by calling the new method twice, to un-assign an then re-assign the aliases.

Reply:
This should still be rebased on #282, because it adds methods like these already, so that this isn't duplicated.

Reply:
Aside of that, the critique of this using aliases from the state instead of aliases from the directory still persists. The aliases from the state are not necessarily correct, and what's more important: local. Getting the list of aliases from the room directory would definitely be better here.

Reply:
> as this is nothing that happens that often and can be done by calling the new method twice, to un-assign an then re-assign the aliases.

I'd still argue that the three steps of a) get list of aliases, b) remove them all from the old room and c) add them all to the new room still have enough inherent complexity that they warrant being moved into a separate method. If not for reusability, then still for readability.

Reply:
I have rebased it onto #282.
#282 only updates the room dir, and the aliases are not announced with "m.room.canonical_alias"

Reply:
And yes. Reading the room dir is possible, but not recommended:
> Note: Clients are recommended not to display this list of aliases prominently as they are not curated, unlike those listed in the m.room.canonical_alias state event.
(https://spec.matrix.org/unstable/client-server-api/#get_matrixclientv3roomsroomidaliases)

Reply:
Well, that's not about reading the room aliases, that's about displaying them. Maybe @poljar has any opinion on this? I'd be in favor of still migrating all local aliases. We should still migrate the state event too, but we need to filter that down to the local aliases anyway, because we can't migrate non-local aliases.

Reply:
Yeah, fair point. That would solve the problem with non-local aliases.
And moving the block into another (private?) method for readability is reasonable. But I want to finish the discussion on the implementation before restructuring the code again.

Reply:
I see no harm in exposing that method to the public